core/primitives/src/block_body.rs

```diff
@@ -17,61 +19,104 @@ pub struct BlockBodyV1 {
 }
 
 impl BlockBodyV1 {
+    // Required as we have raw reference to BlockBodyV1 in BlockV3
     pub fn compute_hash(&self) -> CryptoHash {
         CryptoHash::hash_borsh(self)
     }
 }
 
+/// V1 -> V2: added chunk_endorsements from chunk_validators
+#[derive(BorshSerialize, BorshDeserialize, Debug, Clone, Eq, PartialEq)]
+pub struct BlockBodyV2 {
+    pub chunks: Vec<ShardChunkHeader>,
+    pub challenges: Challenges,
+
+    // Data to confirm the correctness of randomness beacon output
+    pub vrf_value: Value,
+    pub vrf_proof: Proof,
+
+    // Chunk endorsements
+    // These are structured as a vector of Signatures from all ordered chunk_validators
+    // for each shard got from fn get_ordered_chunk_validators
+    // chunk_endorsements[shard_id][chunk_validator_index] is the signature (if present).
+    // If the chunk_validator did not endorse the chunk, the signature is None.
+    pub chunk_endorsements: Vec<Vec<Option<Box<Signature>>>>,
```

Comment:
Should we also specify what happens for chunks that are missing and thus copied from the previous block? Should we copy the endorsements too or should that entry of the outer vector be an empty vector?

Reply:
For now going with copying the old chunk endorsements from previous block.